iceoryx_examples/ara_com/ara/runtime.hpp

```diff
@@ -0,0 +1,171 @@
+    {
+        static Runtime runtime;
+        return runtime;
+    }
+
+    ara::com::ServiceHandleContainer<com::ProxyHandleType>
+    FindService(ara::com::ServiceIdentifier& serviceIdentifier,
+                ara::com::InstanceIdentifier& instanceIdentifier) noexcept
+    {
+        ara::com::ServiceHandleContainer<com::ProxyHandleType> iceoryxServiceContainer;
+
+        m_discovery.findService(
+            serviceIdentifier,
+            instanceIdentifier,
+            iox::cxx::nullopt,
+            [&](auto& service) {
+                iceoryxServiceContainer.push_back({service.getEventIDString(), service.getInstanceIDString()});
+            },
+            iox::popo::MessagingPattern::PUB_SUB);
+
+        m_discovery.findService(
+            serviceIdentifier,
+            instanceIdentifier,
+            iox::cxx::nullopt,
+            [&](auto& service) {
+                iceoryxServiceContainer.push_back({service.getEventIDString(), service.getInstanceIDString()});
+            },
+            iox::popo::MessagingPattern::REQ_RES);
+
+        // We need to make sure that all three internal services representing 'MinimalSkeleton' are available
+        ara::com::ServiceHandleContainer<com::ProxyHandleType> araServiceContainer;
+        if (verifyThatServiceIsComplete(iceoryxServiceContainer))
+        {
+            araServiceContainer.push_back({serviceIdentifier, instanceIdentifier});
+        }
+
+        return araServiceContainer;
+    }
+
+    ara::com::FindServiceHandle StartFindService(ara::com::FindServiceHandler<com::ProxyHandleType> handler,
+                                                 ara::com::ServiceIdentifier& serviceIdentifier,
+                                                 ara::com::InstanceIdentifier& instanceIdentifier) noexcept
+    {
+        /// @todo #1332 Are duplicate entries allowed?
+        m_callbacks.push_back({handler, {serviceIdentifier, instanceIdentifier}});
+
+        if (m_callbacks.size() == 1)
+        {
+            auto invoker = iox::popo::createNotificationCallback(invokeCallback, *this);
+            m_listener.attachEvent(m_discovery, iox::runtime::ServiceDiscoveryEvent::SERVICE_REGISTRY_CHANGED, invoker)
+                .or_else([](auto) {
+                    std::cerr << "unable to attach discovery" << std::endl;
+                    std::exit(EXIT_FAILURE);
+                });
+        }
+
+        return ara::com::FindServiceHandle({serviceIdentifier, instanceIdentifier});
+    }
+
+    void StopFindService(ara::com::FindServiceHandle handle) noexcept
+    {
+        auto iter = m_callbacks.begin();
+        for (; iter != m_callbacks.end(); iter++)
+        {
+            if (iter->second.GetServiceId() == handle.GetServiceId()
+                && iter->second.GetInstanceId() == handle.GetInstanceId())
+            {
+                break;
+            }
+        }
+        if (iter != m_callbacks.end())
+        {
+            m_callbacks.erase(iter);
```

Comment:
This is bad as we erase in the middle of a vector in general. Can we please use the right data structure for the job? It is possible to create one on top of vector (slightly suboptimal, but still better).

This will also make finding an entry based on the handle (=index) faster. This is what I was sketching in my original PR on how to mimic Autosar discovery but did not yet implement it.

Reply:
@MatthiasKillat If there's not reason beside performance I wouldn't exchange the data structure as this example has the purpose of just showing how it could be done.

Reply:
I added a comment.